client/host_queue.go

```diff
@@ -0,0 +1,283 @@
+	writeRequestArrayPool writeRequestArrayPool,
+	opts m3db.ClientOptions,
+) hostQueue {
+	size := opts.GetHostQueueOpsFlushSize()
+
+	q := &queue{
+		host:                  host,
+		connPool:              newConnectionPool(host, opts),
+		writeBatchRequestPool: writeBatchRequestPool,
+		writeRequestArrayPool: writeRequestArrayPool,
+		size: size,
+		ops:  make([]op, 0, size),
+		// NB(r): specifically use non-buffered queues for draining
+		// to ensure safe swapping of ops in and out
+		drainIn:      make(chan []op),
+		drainRecycle: make(chan []op, maxRecycleOpsFlush),
+		closed:       false,
+	}
+
+	// Continually drain the queue until closed
+	go q.drain()
+
+	flushInterval := opts.GetHostQueueOpsFlushInterval()
+	if flushInterval > 0 {
+		// Continually flush the queue at given interval if set
+		go q.flushEvery(flushInterval)
+	}
+
+	return q
+}
+
+func (q *queue) flushEvery(interval time.Duration) {
+	for {
+		q.RLock()
+		closed := q.closed
+		q.RUnlock()
+		if closed {
+			return
+		}
+
+		time.Sleep(interval)
+
+		q.Lock()
+		if q.closed {
+			q.Unlock()
+			return
+		}
+		q.flushWithLock()
+		q.Unlock()
+	}
+}
+
+func (q *queue) flushWithLock() {
+	// Pass the current ops to drain
+	q.drainIn <- q.ops
+
+	// Reset ops
+	q.ops = q.getOps()
+}
+
+func (q *queue) getOps() []op {
+	// Retrieve recycled ops or create new
+	var newOps []op
+	select {
+	case newOps = <-q.drainRecycle:
+	default:
+		newOps = make([]op, 0, q.size)
+	}
+	newOps = newOps[:0]
+	return newOps
+}
+
+func (q *queue) putOps(ops []op) {
+	// Recycle ops
+	select {
+	case q.drainRecycle <- ops:
+	default:
+	}
+}
+
+func (q *queue) drain() {
+	wgAll := &sync.WaitGroup{}
+	for {
+		ops := <-q.drainIn
+
+		var (
+			currWriteOps      []op
+			currWriteRequests []*rpc.WriteRequest
+			writeBatchSize    = q.opts.GetWriteBatchSize()
+			opsLen            = len(ops)
+		)
+		for i := 0; i < opsLen; i++ {
+			switch v := ops[i].(type) {
+			case *writeOp:
+				if currWriteOps == nil {
+					currWriteOps = q.getOps()
+					currWriteRequests = q.writeRequestArrayPool.Get()
+				}
+
+				currWriteOps = append(currWriteOps, ops[i])
+				currWriteRequests = append(currWriteRequests, &v.request)
+
+				if len(currWriteOps) == writeBatchSize {
+					q.asyncWrite(wgAll, currWriteOps, currWriteRequests)
+					currWriteOps = nil
+					currWriteRequests = nil
+				}
+			default:
+				completionFn := ops[i].GetCompletionFn()
+				completionFn(nil, errUnknownOperation)
+			}
+		}
+
+		// If any outstanding write ops, async write
+		if len(currWriteOps) > 0 {
+			q.asyncWrite(wgAll, currWriteOps, currWriteRequests)
+		}
+
+		q.putOps(ops)
+
+		q.RLock()
+		closed := q.closed
+		q.RUnlock()
+
+		if closed {
+			// Final drain, close the connection pool after all requests done
+			wgAll.Wait()
+			q.connPool.Close()
+			return
+		}
+	}
+}
+
+func (q *queue) asyncWrite(wg *sync.WaitGroup, ops []op, elems []*rpc.WriteRequest) {
+	wg.Add(1)
+	go func() {
```

Comment:
I thought about that, we don't currently have any worker pool's in the project just yet.  I'll add some in a bit and put a TODO here perhaps for now?